heat/ml/cluster/kmeans.py

```diff
@@ -4,51 +4,351 @@
+                        proc = p
+                    xi = ht.zeros(X.shape[1])
+                    if X.comm.rank == proc:
+                        idx = sample - displ[proc]
+                        xi = ht.array(X.lloc[idx, :, 0])
+                    xi.comm.Bcast(xi, root=proc)
+                    centroids[0, :, i] = xi
+
+            else:
+                raise NotImplementedError('Not implemented for other splitting-axes')
+
+            self._cluster_centers = centroids
+
+        # directly passed centroids
+        elif isinstance(self.init, ht.DNDarray):
+            if len(self.init.shape) != 2:
+                raise ValueError('passed centroids need to be two-dimensional, but are {}'.format(len(self.init)))
+            if self.init.shape[0] != self.n_clusters or self.init.shape[1] != X.shape[1]:
+                raise ValueError('passed centroids do not match cluster count or data shape')
+            self._cluster_centers = self.init.resplit(None).T.expand_dims(axis=0)
+
+        # kmeans++, smart centroid guessing
+        elif self.init == 'kmeans++':
+            if (X.split == None) or (X.split == 0):
+                centroids = ht.empty((1, X.shape[1], self.n_clusters), split=None)
+                sample = ht.random.randint(0, X.shape[0] - 1).item()
+                _, displ, _ = X.comm.counts_displs_shape(shape=X.shape, axis=0)
+                proc = 0
+                for p in range(X.comm.size):
+                    if displ[p] > sample:
+                        break
+                    proc = p
+                x0 = ht.zeros(X.shape[1])
+                if X.comm.rank == 0: print("".format(proc))
+                if X.comm.rank == proc:
+                    idx = sample - displ[proc]
+                    x0 = ht.array(X.lloc[idx, :, 0])
+                x0.comm.Bcast(x0, root=proc)
+                centroids[0, :, 0] = x0
+
+                for i in range(1, self.n_clusters):
+                    distances = ((X - centroids[:, :, :i]) ** 2).sum(axis=1, keepdim=True)
+                    D2 = distances.min(axis=2)
+                    D2.resplit_(axis=None)
+                    D2 = D2.squeeze()
+                    prob = D2 / D2.sum()
+                    x = ht.random.rand().item()
+                    sample = 0
+                    sum = 0
+                    for j in range(len(prob)):
+                        if sum > x:
+                            break
+                        sum += prob[j].item()
+                        sample = j
+                    proc = 0
+                    for p in range(X.comm.size):
+                        if displ[p] > sample:
+                            break
+                        proc = p
+                    xi = ht.zeros(X.shape[1])
+                    if X.comm.rank == proc:
+                        idx = sample - displ[proc]
+                        xi = ht.array(X.lloc[idx, :, 0])
+                    xi.comm.Bcast(xi, root=proc)
```

Comment:
Same here

Reply:
Nope, because the next iteration needs to have all the previously determined centroids to calculate distances"